{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notebook 12: HG002 Contamination Fix For Genbank<a class=\"tocSkip\">\n",
    "\n",
    "**HG002's maternal assembly needs to have a contig dropped and it needs to have some (more) adapter sequence masked. Genbank has asked us to do this (they don't mask adapters).*\n",
    "    \n",
    "    \n",
    "**The steps that we will take are:**\n",
    "1. Import Statements & Global Variable Definitions\n",
    "2. Download HG002 Maternal Files\n",
    "3. Drop EBV Contig\n",
    "4. Mask Adapter Sequence\n",
    "5. Fix Headers\n",
    "6. Copy Final FASTA To Bucket"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import Statements & Global Variable Definitions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Python packages\n",
    "----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture \n",
    "import terra_notebook_utils as tnu\n",
    "import terra_pandas as tp\n",
    "import os\n",
    "import io\n",
    "import gzip\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq, Alphabet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set Environment Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Billing project: human-pangenome-ucsc\n",
      "Workspace: HPRC_Reassembly\n",
      "Workspace storage bucket: gs://fc-0c2122a8-6725-4199-b90e-828ab006078f/\n"
     ]
    }
   ],
   "source": [
    "# Get the Google billing project name and workspace name\n",
    "PROJECT = os.environ['WORKSPACE_NAMESPACE']\n",
    "WORKSPACE = os.path.basename(os.path.dirname(os.getcwd()))\n",
    "bucket = os.environ['WORKSPACE_BUCKET'] + \"/\"\n",
    "\n",
    "# Verify that we've captured the environment variables\n",
    "print(\"Billing project: \" + PROJECT)\n",
    "print(\"Workspace: \" + WORKSPACE)\n",
    "print(\"Workspace storage bucket: \" + bucket)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Download HG002 Maternal Files\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘HG002_Genbank’: File exists\n",
      "/home/jupyter-user/notebooks/HPRC_Reassembly/edit/HG002_Genbank\n"
     ]
    }
   ],
   "source": [
    "! mkdir HG002_Genbank\n",
    "%cd HG002_Genbank"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download partially fixed assembly + remaining contamination"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2021-05-27 16:39:23--  https://submit.ncbi.nlm.nih.gov/api/2.0/files/8jtviypy/foreigncontaminationmodified_hg002_maternal_f1_assembly_v2_1_genbank.zip/?format=attachment\n",
      "Resolving submit.ncbi.nlm.nih.gov (submit.ncbi.nlm.nih.gov)... 130.14.29.113, 2607:f220:41e:4290::113\n",
      "Connecting to submit.ncbi.nlm.nih.gov (submit.ncbi.nlm.nih.gov)|130.14.29.113|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 885141678 (844M) [application/zip]\n",
      "Saving to: ‘foreigncontaminationmodified_hg002_maternal_f1_assembly_v2_1_genbank.zip’\n",
      "\n",
      "foreigncontaminatio 100%[===================>] 844.14M  32.0MB/s    in 22s     \n",
      "\n",
      "2021-05-27 16:39:46 (38.2 MB/s) - ‘foreigncontaminationmodified_hg002_maternal_f1_assembly_v2_1_genbank.zip’ saved [885141678/885141678]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! wget \\\n",
    "    https://submit.ncbi.nlm.nih.gov/api/2.0/files/8jtviypy/foreigncontaminationmodified_hg002_maternal_f1_assembly_v2_1_genbank.zip/?format=attachment \\\n",
    "    --output-document foreigncontaminationmodified_hg002_maternal_f1_assembly_v2_1_genbank.zip "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2021-05-27 16:39:54--  https://submit.ncbi.nlm.nih.gov/api/2.0/files/uuwykxej/remainingcontamination_hg002_maternal_f1_assembly_v2_1_genbank.txt/?format=attachment\n",
      "Resolving submit.ncbi.nlm.nih.gov (submit.ncbi.nlm.nih.gov)... 130.14.29.113, 2607:f220:41e:4290::113\n",
      "Connecting to submit.ncbi.nlm.nih.gov (submit.ncbi.nlm.nih.gov)|130.14.29.113|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 4640 (4.5K) [text/plain]\n",
      "Saving to: ‘remainingcontamination_hg002_maternal_f1_assembly_v2_1_genbank.txt’\n",
      "\n",
      "remainingcontaminat 100%[===================>]   4.53K  --.-KB/s    in 0.001s  \n",
      "\n",
      "2021-05-27 16:39:54 (4.28 MB/s) - ‘remainingcontamination_hg002_maternal_f1_assembly_v2_1_genbank.txt’ saved [4640/4640]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! wget \\\n",
    "    https://submit.ncbi.nlm.nih.gov/api/2.0/files/uuwykxej/remainingcontamination_hg002_maternal_f1_assembly_v2_1_genbank.txt/?format=attachment \\\n",
    "    --output-document remainingcontamination_hg002_maternal_f1_assembly_v2_1_genbank.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Archive:  foreigncontaminationmodified_hg002_maternal_f1_assembly_v2_1_genbank.zip\n",
      "  inflating: fasta/HG002_maternal_f1_assembly_v2_1_genbank00000000.fsa  \n"
     ]
    }
   ],
   "source": [
    "! unzip foreigncontaminationmodified_hg002_maternal_f1_assembly_v2_1_genbank.zip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 2.9G\r\n",
      "-rw-rw-r-- 1 jupyter-user users 2.9G May 11 11:40 HG002_maternal_f1_assembly_v2_1_genbank00000000.fsa\r\n"
     ]
    }
   ],
   "source": [
    "! ls -lh fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Take a look at what we need to fix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(9 split spans with locations to mask/trim)\r\n",
      "\r\n",
      "Trim:\r\n",
      "Sequence name, length, span(s), apparent source\r\n",
      "HG002#2#h2tg000001l\t110635364\t38657030..38657078\tadaptor:NGB00972.1-not_cleaned\r\n",
      "HG002#2#h2tg000004l\t111658246\t93366436..93366567\tmitochondrion-not_cleaned\r\n",
      "HG002#2#h2tg000037l\t47503194\t32469744..32470038\tmitochondrion-not_cleaned\r\n",
      "HG002#2#h2tg000051l\t86491595\t72699959..72700004\tadaptor:NGB00972.1-not_cleaned\r\n",
      "HG002#2#h2tg000077l\t57225567\t4930882..4931137\tmitochondrion-not_cleaned\r\n",
      "HG002#2#h2tg000115l\t55119985\t32266994..32267188\tmitochondrion-not_cleaned\r\n",
      "HG002#2#h2tg000128l\t4104114\t934938..935807\tmitochondrion-not_cleaned\r\n",
      "HG002#2#h2tg000180l\t870869\t64581..70422\tmitochondrion-not_cleaned\r\n",
      "HG002#2#h2tg000535l\t45111\t11417..45110\tHuman gammaherpesvirus 4-not_cleaned\r\n",
      "\r\n",
      "\r\n"
     ]
    }
   ],
   "source": [
    "! tail -n 15 *.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**So we need to mask these regions:**<br>\n",
    "HG002#2#h2tg000001l\t110635364\t38657030..38657078\tadaptor:NGB00972.1-not_cleaned <br>\n",
    "HG002#2#h2tg000051l\t86491595\t72699959..72700004\tadaptor:NGB00972.1-not_cleaned <br>\n",
    "\n",
    "**And we need to remove this contig**<br>\n",
    "HG002#2#h2tg000535l\t45111\t11417..45110\tHuman gammaherpesvirus 4-not_cleaned\n",
    "\n",
    "*Note that I have already blasted 1-10000 of this contig and found it to also be EBV, so Shelby and I agreed to drop the whole thing...*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Drop EBV Contig\n",
    "**HG002#2#h2tg000535l**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/jupyter-user/notebooks/HPRC_Reassembly/edit/HG002_Genbank/fasta\n"
     ]
    }
   ],
   "source": [
    "%cd fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Drop Contig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "! samtools faidx HG002_maternal_f1_assembly_v2_1_genbank00000000.fsa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "! mkdir dropped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "! cat HG002_maternal_f1_assembly_v2_1_genbank00000000.fsa.fai \\\n",
    "    | cut -f1 | grep -v 'HG002#2#h2tg000535l' \\\n",
    "    > dropped/HG002.mat.goodcontigs.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "445\n",
      "446\n"
     ]
    }
   ],
   "source": [
    "## Check that it worked\n",
    "! cat dropped/HG002.mat.goodcontigs.txt | wc -l\n",
    "! cat HG002_maternal_f1_assembly_v2_1_genbank00000000.fsa.fai | wc -l"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Actually drop the contig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "! samtools faidx \\\n",
    "    HG002_maternal_f1_assembly_v2_1_genbank00000000.fsa \\\n",
    "    `cat dropped/HG002.mat.goodcontigs.txt` \\\n",
    "    > dropped/HG002.mat.dropped.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check Our Work"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.06065e+09\r\n"
     ]
    }
   ],
   "source": [
    "! grep -v \">\" HG002_maternal_f1_assembly_v2_1_genbank00000000.fsa | wc | awk '{print $3-$1}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.06061e+09\r\n"
     ]
    }
   ],
   "source": [
    "! grep -v \">\" dropped/HG002.mat.dropped.fa | wc | awk '{print $3-$1}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "! samtools faidx dropped/HG002.mat.dropped.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "445\r\n"
     ]
    }
   ],
   "source": [
    "! cat dropped/HG002.mat.dropped.fa.fai | wc -l"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Size and number of contigs looks good**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mask Adapter Sequence\n",
    "## Pull Adapter Sequence To Check\n",
    "HG002#2#h2tg000001l 110635364 38657030..38657078 adaptor:NGB00972.1-not_cleaned <br>\n",
    "HG002#2#h2tg000051l 86491595 72699959..72700004 adaptor:NGB00972.1-not_cleaned <br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">lcl|HG002#2#h2tg000001l:38657030-38657078\r\n",
      "ATCTCTCTCAAACAACAACAACGGGAGGAGAGGAAAAGAGAGAGATAAC\r\n"
     ]
    }
   ],
   "source": [
    "! samtools faidx \\\n",
    "    dropped/HG002.mat.dropped.fa \\\n",
    "    'lcl|HG002#2#h2tg000001l:38657030-38657078'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Compare against SMRTbell Dimer\n",
    "\n",
    "# TCTCTCAACAACAACAACGGAGG-AGGAGGAAAAGAGAGAGATATCTCTCTCAACAACAACAACGGAGGAGGAGGAAAAGAGAGAGAT\n",
    "# ATCTCTCTCAAACAACAACAACGGGAGGAGAGGAAAAGAGAGAGATAAC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">lcl|HG002#2#h2tg000051l:72699959-72700004\r\n",
      "ATCTCTCTCAACAACATCAACGGAGGAGGAGGGAAAAGAGAGAGAT\r\n"
     ]
    }
   ],
   "source": [
    "! samtools faidx \\\n",
    "    dropped/HG002.mat.dropped.fa \\\n",
    "    'lcl|HG002#2#h2tg000051l:72699959-72700004'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Compare against SMRTbell dimer\n",
    "\n",
    "#  TCTCT--CAACAACAACAACGGAGGAGGAGG-AAAAGAGAGAGATATCTCTCTCAACAACAACAACGGAGGAGGAGGAAAAGAGAGAGAT\n",
    "# ATCTCTCTCAACAACATCAACGGAGGAGGAGGGAAAAGAGAGAGAT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘mask’: File exists\r\n"
     ]
    }
   ],
   "source": [
    "! mkdir mask\n",
    "\n",
    "! echo 'lcl|HG002#2#h2tg000001l\\t38657030\\t38657078\\tadaptor:NGB00972.1-not_cleaned' > mask/HG002_mat_mask.bed\n",
    "! echo 'lcl|HG002#2#h2tg000051l\\t72699959\\t72700004\\tadaptor:NGB00972.1-not_cleaned' >> mask/HG002_mat_mask.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lcl|HG002#2#h2tg000001l\t38657030\t38657078\tadaptor:NGB00972.1-not_cleaned\r\n",
      "lcl|HG002#2#h2tg000051l\t72699959\t72700004\tadaptor:NGB00972.1-not_cleaned\r\n"
     ]
    }
   ],
   "source": [
    "! cat mask/HG002_mat_mask.bed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mask Adapters"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Install BedTools**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2021-05-27 22:50:22--  https://github.com/arq5x/bedtools2/releases/download/v2.30.0/bedtools.static.binary\n",
      "Resolving github.com (github.com)... 140.82.112.4\n",
      "Connecting to github.com (github.com)|140.82.112.4|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 Found\n",
      "Location: https://github-releases.githubusercontent.com/15059334/c633cf80-61f8-11eb-92ef-18b90dff37e2?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20210527%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210527T225022Z&X-Amz-Expires=300&X-Amz-Signature=559be6e2cffed8c5d40d717177867c4ef83405fbb3ba618dd61d87d93786c4e4&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=15059334&response-content-disposition=attachment%3B%20filename%3Dbedtools.static.binary&response-content-type=application%2Foctet-stream [following]\n",
      "--2021-05-27 22:50:22--  https://github-releases.githubusercontent.com/15059334/c633cf80-61f8-11eb-92ef-18b90dff37e2?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20210527%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210527T225022Z&X-Amz-Expires=300&X-Amz-Signature=559be6e2cffed8c5d40d717177867c4ef83405fbb3ba618dd61d87d93786c4e4&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=15059334&response-content-disposition=attachment%3B%20filename%3Dbedtools.static.binary&response-content-type=application%2Foctet-stream\n",
      "Resolving github-releases.githubusercontent.com (github-releases.githubusercontent.com)... 185.199.108.154, 185.199.109.154, 185.199.110.154, ...\n",
      "Connecting to github-releases.githubusercontent.com (github-releases.githubusercontent.com)|185.199.108.154|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 41993576 (40M) [application/octet-stream]\n",
      "Saving to: ‘bedtools.static.binary’\n",
      "\n",
      "bedtools.static.bin 100%[===================>]  40.05M  39.5MB/s    in 1.0s    \n",
      "\n",
      "2021-05-27 22:50:23 (39.5 MB/s) - ‘bedtools.static.binary’ saved [41993576/41993576]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! wget https://github.com/arq5x/bedtools2/releases/download/v2.30.0/bedtools.static.binary\n",
    "    \n",
    "! mv bedtools.static.binary bedtools\n",
    "! chmod a+x bedtools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "! ./bedtools maskfasta \\\n",
    "    -fi dropped/HG002.mat.dropped.fa \\\n",
    "    -bed mask/HG002_mat_mask.bed \\\n",
    "    -fo mask/HG002.mat.dropped.masked.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check What We Did\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "419\n",
      "326\n",
      "326\n"
     ]
    }
   ],
   "source": [
    "! cat mask/HG002.mat.dropped.masked.fa | grep -v 'lcl|' | grep -o \"N\" | wc -l\n",
    "! cat dropped/HG002.mat.dropped.fa | grep -v 'lcl|' | grep -o \"N\" | wc -l\n",
    "! cat HG002_maternal_f1_assembly_v2_1_genbank00000000.fsa | grep -v 'lcl|' | grep -o \"N\" | wc -l"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Take a look at the headers**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">lcl|HG002#2#h2tg000531l\r\n",
      ">lcl|HG002#2#h2tg000532l\r\n",
      ">lcl|HG002#2#h2tg000533l\r\n",
      ">lcl|HG002#2#h2tg000536l\r\n",
      ">lcl|HG002#2#h2tg000537l\r\n",
      ">lcl|HG002#2#h2tg000538l\r\n",
      ">lcl|HG002#2#h2tg000541c\r\n",
      ">lcl|HG002#2#h2tg000542l\r\n",
      ">lcl|HG002#2#h2tg000544l\r\n",
      ">lcl|HG002#2#MT\r\n"
     ]
    }
   ],
   "source": [
    "! cat mask/HG002.mat.dropped.masked.fa | grep '>' | tail"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">lcl|HG002#2#h2tg000531l\r\n",
      ">lcl|HG002#2#h2tg000532l\r\n",
      ">lcl|HG002#2#h2tg000533l\r\n",
      ">lcl|HG002#2#h2tg000536l\r\n",
      ">lcl|HG002#2#h2tg000537l\r\n",
      ">lcl|HG002#2#h2tg000538l\r\n",
      ">lcl|HG002#2#h2tg000541c\r\n",
      ">lcl|HG002#2#h2tg000542l\r\n",
      ">lcl|HG002#2#h2tg000544l\r\n",
      ">lcl|HG002#2#MT\r\n"
     ]
    }
   ],
   "source": [
    "! cat dropped/HG002.mat.dropped.fa | grep '>' | tail"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">lcl|HG002#2#h2tg000532l Homo sapiens isolate NA24385\r\n",
      ">lcl|HG002#2#h2tg000533l Homo sapiens isolate NA24385\r\n",
      ">lcl|HG002#2#h2tg000535l Homo sapiens isolate NA24385\r\n",
      ">lcl|HG002#2#h2tg000536l Homo sapiens isolate NA24385\r\n",
      ">lcl|HG002#2#h2tg000537l Homo sapiens isolate NA24385\r\n",
      ">lcl|HG002#2#h2tg000538l Homo sapiens isolate NA24385\r\n",
      ">lcl|HG002#2#h2tg000541c Homo sapiens isolate NA24385\r\n",
      ">lcl|HG002#2#h2tg000542l Homo sapiens isolate NA24385\r\n",
      ">lcl|HG002#2#h2tg000544l Homo sapiens isolate NA24385\r\n",
      ">lcl|HG002#2#MT Homo sapiens isolate NA24385 mitochondrion\r\n"
     ]
    }
   ],
   "source": [
    "! cat HG002_maternal_f1_assembly_v2_1_genbank00000000.fsa | grep '>' | tail"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**So the headers need to be fixed**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fix Headers\n",
    "**Need to add \"Homo sapiens isolate NA24385\" to the end of all headers, and need to add \"mitochondrion\" to MT header.**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fix By Adding Sample Info And MT Signifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "! mkdir rename_headers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add sample info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "! sed '/>lcl/s/$/ Homo sapiens isolate NA24385/' \\\n",
    "    mask/HG002.mat.dropped.masked.fa \\\n",
    "    > rename_headers/HG002.mat.dropped.masked.renamed_pt1.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add Mito ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "! sed '/MT/s/$/ mitochondrion/' \\\n",
    "    rename_headers/HG002.mat.dropped.masked.renamed_pt1.fa \\\n",
    "    > rename_headers/HG002.mat.dropped.masked.renamed_pt2.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check Everythign Again On Final File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">lcl|HG002#2#h2tg000531l Homo sapiens isolate NA24385\r\n",
      ">lcl|HG002#2#h2tg000532l Homo sapiens isolate NA24385\r\n",
      ">lcl|HG002#2#h2tg000533l Homo sapiens isolate NA24385\r\n",
      ">lcl|HG002#2#h2tg000536l Homo sapiens isolate NA24385\r\n",
      ">lcl|HG002#2#h2tg000537l Homo sapiens isolate NA24385\r\n",
      ">lcl|HG002#2#h2tg000538l Homo sapiens isolate NA24385\r\n",
      ">lcl|HG002#2#h2tg000541c Homo sapiens isolate NA24385\r\n",
      ">lcl|HG002#2#h2tg000542l Homo sapiens isolate NA24385\r\n",
      ">lcl|HG002#2#h2tg000544l Homo sapiens isolate NA24385\r\n",
      ">lcl|HG002#2#MT Homo sapiens isolate NA24385 mitochondrion\r\n"
     ]
    }
   ],
   "source": [
    "! cat rename_headers/HG002.mat.dropped.masked.renamed_pt2.fa | grep \">\" | tail"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "445\r\n"
     ]
    }
   ],
   "source": [
    "! cat rename_headers/HG002.mat.dropped.masked.renamed_pt2.fa | grep \"Homo sapiens isolate NA24385\" | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "419\r\n"
     ]
    }
   ],
   "source": [
    "! cat rename_headers/HG002.mat.dropped.masked.renamed_pt2.fa | grep -v 'lcl|' | grep -o \"N\" | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.06061e+09\r\n"
     ]
    }
   ],
   "source": [
    "! grep -v \">\" rename_headers/HG002.mat.dropped.masked.renamed_pt2.fa | wc | awk '{print $3-$1}'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Copy Final FASTA To Bucket"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "! mkdir final_fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Copy Over Maternal (Changed) File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "! gzip rename_headers/HG002.mat.dropped.masked.renamed_pt2.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "! cp \\\n",
    "    rename_headers/HG002.mat.dropped.masked.renamed_pt2.fa.gz \\\n",
    "    final_fasta/HG002.mat.dropped.masked.renamed_pt2.fa.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://final_fasta/HG002.mat.dropped.masked.renamed_pt2.fa.gz [Content-Type=application/octet-stream]...\n",
      "==> NOTE: You are uploading one or more large file(s), which would run          \n",
      "significantly faster if you enable parallel composite uploads. This\n",
      "feature can be enabled by editing the\n",
      "\"parallel_composite_upload_threshold\" value in your .boto\n",
      "configuration file. However, note that if you do this large files will\n",
      "be uploaded as `composite objects\n",
      "<https://cloud.google.com/storage/docs/composite-objects>`_,which\n",
      "means that any user who downloads such objects will need to have a\n",
      "compiled crcmod installed (see \"gsutil help crcmod\"). This is because\n",
      "without a compiled crcmod, computing checksums on composite objects is\n",
      "so slow that gsutil disables downloads of composite objects.\n",
      "\n",
      "- [1 files][851.5 MiB/851.5 MiB]   84.5 MiB/s                                   \n",
      "Operation completed over 1 objects/851.5 MiB.                                    \n"
     ]
    }
   ],
   "source": [
    "! gsutil cp \\\n",
    "    final_fasta/HG002.mat.dropped.masked.renamed_pt2.fa.gz \\\n",
    "    {bucket}fix_hg002_genbank/HG002.mat.dropped.masked.renamed_pt2.fa.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Copy Over Paternal (Unchanged) File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2021-06-01 20:47:48--  https://submit.ncbi.nlm.nih.gov/api/2.0/files/hkf9hszi/foreigncontaminationmodified_hg002_paternal_f1_assembly_v2_1.zip/?format=attachment\n",
      "Resolving submit.ncbi.nlm.nih.gov (submit.ncbi.nlm.nih.gov)... 130.14.29.113, 2607:f220:41e:4290::113\n",
      "Connecting to submit.ncbi.nlm.nih.gov (submit.ncbi.nlm.nih.gov)|130.14.29.113|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 849361683 (810M) [application/zip]\n",
      "Saving to: ‘final_fasta/foreigncontaminationmodified_hg002_paternal_f1_assembly_v2_1.zip’\n",
      "\n",
      "final_fasta/foreign 100%[===================>] 810.01M  23.8MB/s    in 27s     \n",
      "\n",
      "2021-06-01 20:48:16 (30.1 MB/s) - ‘final_fasta/foreigncontaminationmodified_hg002_paternal_f1_assembly_v2_1.zip’ saved [849361683/849361683]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! wget \\\n",
    "    https://submit.ncbi.nlm.nih.gov/api/2.0/files/hkf9hszi/foreigncontaminationmodified_hg002_paternal_f1_assembly_v2_1.zip/?format=attachment \\\n",
    "    --output-document final_fasta/foreigncontaminationmodified_hg002_paternal_f1_assembly_v2_1.zip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Archive:  final_fasta/foreigncontaminationmodified_hg002_paternal_f1_assembly_v2_1.zip\n",
      "  inflating: fasta/HG002_paternal_f1_assembly_v2_100000000.fsa  \n"
     ]
    }
   ],
   "source": [
    "! unzip final_fasta/foreigncontaminationmodified_hg002_paternal_f1_assembly_v2_1.zip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "! cp \\\n",
    "    fasta/HG002_paternal_f1_assembly_v2_100000000.fsa \\\n",
    "    final_fasta/HG002_paternal_f1_assembly_v2_100000000.fsa\n",
    "\n",
    "! gzip final_fasta/HG002_paternal_f1_assembly_v2_100000000.fsa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "! rm final_fasta/foreigncontaminationmodified_hg002_paternal_f1_assembly_v2_1.zip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://final_fasta/HG002_paternal_f1_assembly_v2_100000000.fsa.gz [Content-Type=application/octet-stream]...\n",
      "==> NOTE: You are uploading one or more large file(s), which would run          \n",
      "significantly faster if you enable parallel composite uploads. This\n",
      "feature can be enabled by editing the\n",
      "\"parallel_composite_upload_threshold\" value in your .boto\n",
      "configuration file. However, note that if you do this large files will\n",
      "be uploaded as `composite objects\n",
      "<https://cloud.google.com/storage/docs/composite-objects>`_,which\n",
      "means that any user who downloads such objects will need to have a\n",
      "compiled crcmod installed (see \"gsutil help crcmod\"). This is because\n",
      "without a compiled crcmod, computing checksums on composite objects is\n",
      "so slow that gsutil disables downloads of composite objects.\n",
      "\n",
      "| [1 files][810.0 MiB/810.0 MiB]                                                \n",
      "Operation completed over 1 objects/810.0 MiB.                                    \n"
     ]
    }
   ],
   "source": [
    "! gsutil cp \\\n",
    "    final_fasta/HG002_paternal_f1_assembly_v2_100000000.fsa.gz \\\n",
    "    {bucket}fix_hg002_genbank/HG002_paternal_f1_assembly_v2_100000000.fsa.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
